### PR TITLE
Introduction of a Hexagon Cell Grid in the GridMap node

### PR DESCRIPTION
**Pull request by Wasabi-Cheetah**
I've merged the work done by **BSChad** in his fork of Godot.
It Introduces a new Hexagon Grid option for the GridMap node.
Proper accreditation given in each file since I'm just a merger.

# Locations of files changed for Hexagon feature in GridMap node.
redot/core/math/math_defs.h
modules/gridmap/doc_classes/GridMap.xml
modules/gridmap/editor/grid_map_editor_plugin.cpp
modules/gridmap/editor/grid_map_editor_plugin.h
modules/gridmap/grid_map.cpp

Since I'm on Linux, I will do build for Linux and post it on my fork.

Related: https://github.com/godotengine/godot-proposals/issues/4337 and https://github.com/godotengine/godot/pull/85890

### CODE REVIEW

File: core/math/math_defs.h [hunk 1]
```diff
@@ -41,6 +41,7 @@
 
 #define Math_SQRT12 0.7071067811865475244008443621048490
 #define Math_SQRT2 1.4142135623730950488016887242
+#define Math_SQRT3 1.7320508075688772935274463415059
```
Comment (Spartan322) [inline core/math/math_defs.h]: This is a trivial math change, it does not need attribution.

Comment (Wasabi-Cheetah) [reply core/math/math_defs.h]: Okay, thanks for letting me know, wanted to be thorough.

**Review by rlidwka — CHANGES_REQUESTED** [review]
I believe this merge is done completely wrong. If you're using git for the first time, that's understandable I suppose.

Original work is presumably here?
https://github.com/BSChad/godot/commits/feat-grid-map-hexagonal-cells/

How it is now: All commits are from you. Each file is its own commit. You're giving credits in the file contents.

How it should be: Commit metadata should be exactly the same as original. There should be no changes in the headers at all. Attribution should be given in commit messages.

I'd suggest to use `git cherry-pick`.

**Comment by Wasabi-Cheetah** [comment]
> I believe this merge is done completely wrong. If you're using git for the first time, that's understandable I suppose.
> 
> Original work is presumably here? https://github.com/BSChad/godot/commits/feat-grid-map-hexagonal-cells/

Yeah first time, and yes that's the repo.
Learning as I go about all this of course.
I'll look into that for sure.

**Comment by rlidwka** [comment]
also, this is an open PR in Godot (85890), which they want to release as 4.4

just wait until it appears upstream

Redot should be merging old and/or rejected PRs, not the new ones that would appear in Godot anyway

**Comment by Wasabi-Cheetah** [comment]
It's been ready to merge for months 
so it seemed ignored by the higher ups, 
thought I'd take a stab at it and merge it.
Helped me learn about proper merging.

**Comment by rlidwka** [comment]
Okay, here's how you merge stuff:

```sh
git clone git@github.com:Wasabi-Cheetah/redot-engine.git
cd redot-engine
git checkout -b feat-grid-map-hexagonal-cells
git fetch https://github.com/Redot-Engine/redot-engine master
git reset --hard FETCH_HEAD
git fetch https://github.com/BSChad/godot feat-grid-map-hexagonal-cells
git merge FETCH_HEAD
git push git@github.com:Wasabi-Cheetah/redot-engine.git
```

Then compile the engine and verify that all of it works as expected. That's the most important part, without which your PR has no value.

**Comment by Spartan322** [comment]
Also could I get the related Godot PR/proposal for this?

**Comment by Wasabi-Cheetah** [comment]
> Also could I get the related Godot PR/proposal for this?

I opened a discussion using the redot-proposals repo here.
https://github.com/Redot-Engine/redot-proposals/discussions/41

It contains the reference PRs from Godot by BSChad and octetdev2
I don't know if I did that right since I'm new to making proposals.

**Comment by Spartan322** [comment]
Related: godotengine/godot-proposals#4337 and godotengine/godot#85890

**Comment by Spartan322** [comment]
Personally I'd prefer to wait on this as the PR is still being actively developed, its not abandoned nor stalled nor rejected by Godot.

**Comment by Wasabi-Cheetah** [comment]
> Personally I'd prefer to wait on the Godot PR as its still being actively developed, its not abandoned nor stalled nor rejected by Godot, it had another commit just last week and it does not appear to be finished.

Yeah I noticed that, thinking I jumped the gun.
Some things are still being worked after compile.
Idk if to post my compiled build of it since I got it.

File: core/math/math_defs.h [hunk 2]
```diff
@@ -29,6 +26,9 @@
 /* TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE      */
 /* SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.                 */
 /**************************************************************************/
+/*                     Work done by Octetdev2 and BSChad.                 */
+/*                       Merging done by Wasabi_Cheetah.                  */
+/**************************************************************************/
```
Comment (decryptedchaos) [inline core/math/math_defs.h]: About this,  we can see what the rest of the team says. 

But for future references,  this is generally not a good practice to add credits to the preamble,  if we allowed one person to do this everyone would want to do it and then suddenly you have 100 names in the preamble.

Comment (Spartan322) [reply core/math/math_defs.h]: Co-authors are supposed to be added to the commit message anyway, like what I did with 2d1c65b00d0c320f4f5f010b832ed3ce212446a6 for WhalesState, but this won't be merged anyway cause the Godot PR is still WIP.

**Pull request closed without merging**